File: CSE241_HW07/BigramDyn.cpp
//
// Created by sbecerikli on 12/29/16.
//

#include "BigramDyn.h"


namespace SB_Bigram{

    template <class T>
    BigramDyn<T>::BigramDyn(const BigramDyn<T> &other) { //copyconsstructor
        delete[] Grams;
        delete[] GramsCount;
        Grams = new pair<T, T>[other.getSize()];
        GramsCount = new int [other.getSize()];
        for (int i = 0; i < other.getSize(); ++i) {
            Grams[i] = other.Grams[i];
            GramsCount[i] = other.GramsCount[i];
        }
        setSize(other.getSize());
        setCapacity(other.getCapactiy());
    }
    template <class T>
    BigramDyn<T>& BigramDyn<T>::operator=(const BigramDyn<T> &other) { // assigmnment

        if(this != &other){
            delete[] Grams;
            delete[] GramsCount;

            setCapacity(other.getCapactiy());
            setSize(other.getSize());
            Grams = new pair<T, T>[other.getSize()];
            GramsCount = new int [other.getSize()];
            for (int i = 0; i < other.getSize(); ++i) {
                Grams[i] = other.Grams[i];
                GramsCount[i] = other.GramsCount[i];
            }
            return *this;

        }
        return *this;
    }


    template <class T>
    int BigramDyn<T>::readFile(string filename) throw(exception){

        string line;
        ifstream MyFile(filename.c_str());
        cout << "filename : " << filename << endl;
        if (MyFile.is_open()) {
            getline(MyFile, line);
    
            istringstream WhiteSpace(line);
            string SubStr;
            T v1, v2;
            WhiteSpace >> SubStr;
            int index;
            while(WhiteSpace.good()) {
                string tempSubStr = SubStr;
                WhiteSpace >> SubStr;
                stringstream convert1(SubStr);
                stringstream convert2(tempSubStr);
                convert1 >> v1;
                convert2 >> v2;
                
                if (search(pair<T, T>(v2, v1), &index)) {
                    
                    GramsCount[index] += 1;
                }
                else{
                    insert(pair<T, T>(v2, v1), 1);    
                }
                
                //setSize(getSize() + 1);
            }
            MyFile.close();
            setTotalNumGrams(getSize());

        }
        else
        {
            exception e;
            e.cantFind = true;
            throw e;
        }
        return  getTotalNumgrams();
    }

    template <class T>
    bool BigramDyn<T>::search(pair<T, T> p, int* index){
        for (int i = 0; i < getSize(); ++i) {
            if(Grams[i].first == p.first && Grams[i].second == p.second){
                
                *index = i;
                return true;
            }
        }
        return false;
    }
    template <class T>
    void BigramDyn<T>::insert(pair<T, T> p, int count){
        if(getSize() == getCapactiy()){
            pair<T, T>* tempGrams = new pair<T, T>[getCapactiy() + 10];
            int* tempGramsCount = new int[getCapactiy() + 10];
            setCapacity(getCapactiy() + 10);
            for (int i = 0; i < getSize(); ++i) {
                tempGrams[i] = Grams[i];
                tempGramsCount[i] = GramsCount[i];
            }
            tempGrams[getSize()] = p;
            tempGramsCount[getSize()] = count;
            setSize(getSize() + 1);
        }
        else{
            Grams[getSize()] = p;
            GramsCount[getSize()] = count;
            setSize(getSize() + 1);
        }
    }

    template <class T>
    int BigramDyn<T>::numGrams() throw(exception){

        return getTotalNumgrams();
    }
    template <class T>
    int BigramDyn<T>::numOfGrams(T first, T second) throw(exception){

        for (int i = 0; i < getSize(); ++i) {
            if(Grams[i].first == first && Grams[i].second == second){
                return GramsCount[i];
            }
        }


    }
    template<class T>
    pair<T, T> BigramDyn<T>::maxGrams() throw(exception){
        int counts[getSize()];
        for (int i = 0; i < getSize(); ++i) {
            counts[i] = GramsCount[i];
        }
        int temp = 0;
        int index;
        for(int i = 0 ; i < getSize(); ++i)
        {
            if(counts[i]>temp){
               temp = counts[i];
               index = i; 
            }
        }
        return Grams[index];
    }

    template<class T>
     void BigramDyn<T>::print(ostream& out) const{
       sort(Grams, + Grams+getSize());
       for(int i = 0; i < getSize(); ++i){
            out << Grams[i].first << ", " << Grams[i].second << ": " << GramsCount[i] << endl ;
       }
    }

}

File: CSE241_HW08/BigramDyn.java


import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;

/**
 * Created by Seyfullah Becerikli on 29/12/2016.
 */
public class BigramDyn<E extends Comparable<E>> implements Bigram<E> {
    private Pair<E, E>[] Grams;
    private int [] GramsCount;
    private int totalNumGramsX;
    private int numOfGramsX;
    private int size;
    private int capacity;

    public BigramDyn(){

    }

    public BigramDyn(int dataType){
        Grams = new Pair[50];
        GramsCount = new int[50];
        totalNumGramsX = 0;
        numOfGramsX = 0;
        size = 0;
        capacity = 50;
    }

    public BigramDyn(BigramDyn b){
        this.Grams = b.Grams;
        this.GramsCount = b.GramsCount;
        this.totalNumGramsX = b.totalNumGramsX;
        this.numOfGramsX = b.numOfGramsX;
        this.size = b.size;
        this.capacity = b.capacity;
    }

    @Override
    public void readFile(String filename) throws IOException {
        BufferedReader br = null;
        try {
            br = new BufferedReader(new FileReader(filename));
            String line;
            line = br.readLine();
            int index = 0;
            //System.out.println(line);
            String[] ParsedLine = line.split(" ");
            //System.out.println("len: " + ParsedLine.length);
            int i;
            for (i = 0; i < ParsedLine.length - 1; ++i) {
                if(search(new Pair<E, E>((E)ParsedLine[i], (E)ParsedLine[i+1]), index)){
                    GramsCount[index]+=1;
                    System.out.println("girdi");
                }
                else{
                    insert(new Pair<E, E>((E)ParsedLine[i], (E)ParsedLine[i+1]), 1);
                }

            }
            setSize(i);

        }
        finally {
            br.close();
        }
    }
    public boolean search(Pair<E, E> p, int index){
        for (int i = 0; i < getSize(); i++) {
            if(Grams[i].getFirst().compareTo(p.getFirst()) == 0 && Grams[i].getSecond().compareTo(p.getSecond()) == 0){
                index = i;
                return true;
            }
        }
        return false;
    }

    public void insert(Pair<E, E> p, int count){
        if(getSize() == getCapacity()){
            Pair<E, E> tempGrams[] = new Pair[getCapacity() + 10];
            int [] tempGramsCount = new int[getCapacity() + 10];
            setCapacity(getCapacity() + 10);
            for (int i = 0; i < getSize(); i++) {
                tempGrams[i] = Grams[i];
                tempGramsCount[i] = GramsCount[i];
            }
            tempGrams[getSize()] = p;
            tempGramsCount[getSize()] = count;
            setSize(getSize() + 1);
        }
        else{
            Grams[getSize()] = p;
            GramsCount[getSize()] = count;
            setSize(getSize() + 1);
        }
    }

    @Override
    public int numGrams() {
        return getSize();
    }

    @Override
    public int numOfGrams(E first, E second) {
        for (int i = 0; i < getSize()-1; ++i){
            if(Grams[i].getFirst().equals(first.toString()) && Grams[i].getSecond().equals(second.toString())){
                return GramsCount[i];
            }
        }
        return 0;
    }

    public Pair<E, E>[] getGrams() {
        return Grams;
    }

    public void setGrams(Pair<E, E>[] grams) {
        Grams = grams;
    }

    public int[] getGramsCount() {
        return GramsCount;
    }

    public void setGramsCount(int[] gramsCount) {
        GramsCount = gramsCount;
    }

    public int getTotalNumGramsX() {
        return totalNumGramsX;
    }

    public void setTotalNumGramsX(int totalNumGramsX) {
        this.totalNumGramsX = totalNumGramsX;
    }

    public int getNumOfGramsX() {
        return numOfGramsX;
    }

    public void setNumOfGramsX(int numOfGramsX) {
        this.numOfGramsX = numOfGramsX;
    }

    public int getSize() {
        return size;
    }

    public void setSize(int size) {
        this.size = size;
    }

    public int getCapacity() {
        return capacity;
    }

    public void setCapacity(int capacity) {
        this.capacity = capacity;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < getSize()-1; i++) {
            sb.append(Grams[i] + " , " + GramsCount[i] );
        }
        return sb.toString();
    }

}
